rfcs/0039-buffer-physical-layout.md

```diff
@@ -0,0 +1,640 @@
+should be converted to `AllocateNode`.  `LoadNode` and `StoreNode`
+are deprecated.
+
+## Impacted TIR Nodes
+
+- BufferNode
+  - Describes a N-d buffer.  The layout of the buffer may be
+
+- BufferRealizeNode
+  - Realization of a buffer, in logical layout.
+  - For external buffers, serves as an optional annotation.  For
+    internal buffers, results in allocation of memory.
+
+
+- BufferLoadNode/BufferStoreNode
+  - Read/write of a buffer.
+
+  - Change from previous behavior: Will exist throughout the lowering
+    process, and will be passed to the low-level code generators.
+    Transformations that previously created `Load` and `Store` nodes
+    will instead create `BufferLoad` and `BufferStore` nodes with 1-d
+    indices.
+
+
+
+- AllocateNode
+  - Allocation of a buffer, in physical layout.
+
+  - Declares an allocation of a buffer.
+
+  - Change from previous behavior: Previously, `AllocateNode` held the
+    `buffer_var`, datatype, and buffer extents directly. After
+    implementation of this RFC, `AllocateNode` will instead hold the
+    `Buffer` that is to be allocated.
+
+
+- LoadNode/StoreNode
+  - Read/write of a 1-d buffer, given a `Var` pointer to the start of
+    the buffer and a single index.
+
+  - Deprecated, should instead use `BufferLoad` and `BufferStore` with
+    a 1-d index.
+
+
+## Impacted tir Transformations
+
+- `ApplyBufferTransforms`
+  - A new pass that takes as input a TIR graph that may have buffer
+    transformations stored in the `PrimFunc` attributes.  Returns
+    a TIR graph with all buffer transforms applied as specified.
+
+  - Rewrite `indices` in BufferStore/BufferLoad nodes based on the
+    specified transformation.
+
+  - The transformations are stored as a `Map<Var, Array<IndexMap>>` in
+    the `"buffer_layout_transformations"` attribute of a primfunc.
+    All buffers whose `BufferNode::data` is a key in this map should
+    have their physical layout rewritten.  If the array contains
+    multiple transformations, they are applied sequentially.
+
+    A possible structure for the `IndexMap` node is shown
+    below.
+
+    ```
+    class IndexMapNode : public Object {
+    public:
+      /*! \brief Variables representing the indices prior to remapping.
+       *
+       * If initial_index is empty, then final_index should also be
+       * empty, and no mapping is applied.
+       */
+      Array<Var> initial_index;
+
+      /*!
+       * \brief Expressions defining the indices after remapping.
+       *
+       * These expressions should only be in terms of the initial_index,
+       * and must be expressible as a `tvm::arith::IterSumExpr`.  The
+       * mapping from `initial_index` to `final_index` must be injective.
+       *
+       * If final_index is empty, then initial_index should also be
+       * empty, and the map is an identity function.
+       */
+      Array<PrimExpr> final_index;
+    };
+    ```
+
+  - After applying the transformations, the
```

Comment:
Debugging would be done by tracing changes forward during the compilation process, such as with `tvm.ir.instrument` to view the before and after changes made from each lowering pass.  I think this is the preferred direction for debugging, as it also handles existing cases, such as `Prefetch` nodes that are lowered to for-loops that implement the prefetching.

My concern with leaving the transformation attribute defined after usage is that the validity of an IRModule, or even what computation it will produce, then depends on where that IRModule occurs in the lowering process.  Examples of each are below.

* Ambiguously valid: Consider a transformation from NCHW to NCHWc layout, `lambda n,c,h,w: [n, c//4, h, w, c%4]`.  This transformation is only valid to apply on a rank-4 tensor, and produces a rank-5 tensor.  If the annotation is kept in the TIR graph following the transformation, then an invalid graph (incorrectly attempting to apply the transform on a rank-5 tensor) and a valid graph (already having applied the transform onto a rank-4 tensor, producing a rank-5) would have the same representation.
  
* Ambiguous computation: Consider a transformation from NCHW to NHWC layout, `lambda n,c,h,w: [n,h,w,c]`, and the following psuedo-code compute definition.
  
  ```python
  buf.shape = [A,B,C,D]
  buf.layout_transforms = [lambda n,c,h,w: [n,h,w,c]]
  
  for a in range(A):
      for b in range(B):
          acc = 0.0
          for c in range(C):
              for d in range(D):
                  acc += buf[a,b,c,d]
                  
          for c in range(C):
              for d in range(D):
                  buf[a,b,c,d] = buf[a,b,c,d] / acc
  ```
  
  If the `layout_transforms` annotation is left in the TIR graph after being applied, then the above compute definition could be performing normalization along the height/width dimensions if the transformation hasn't yet been applied, or along the width/channel dimensions if the transformation has been applied.